runtime/src/main/java/org/corfudb/runtime/collections/ProtobufIndexer.java

```diff
@@ -106,30 +126,64 @@
         }
     }
 
+    /**
+     * Inspect field descriptor to verify if it contains a 'oneOf' type field.
+     */
+    private OneofDescriptor inspectDescriptorForOneOfFields(FieldDescriptor nestedFieldDescriptor,
+                                                            Map<String, FieldDescriptor> oneOfFieldNameToDescriptor) {
+        OneofDescriptor oneOfDescriptor = nestedFieldDescriptor.getContainingOneof();
+        if (oneOfDescriptor != null) {
+            List<FieldDescriptor> descriptors = oneOfDescriptor.getFields();
+            for (FieldDescriptor fieldDescriptor : descriptors) {
+                oneOfFieldNameToDescriptor.put(fieldDescriptor.getName(), fieldDescriptor);
+            }
+        }
+        return oneOfDescriptor;
+    }
+
+    /**
+     * Verify if message is valid.
+     *
+     * In the case of a 'oneOf' type field, it is valid if the field set is exactly 'fieldName'
+     */
+    private boolean isValidField(OneofDescriptor oneOfDescriptor, Message message, String fieldName) {
+        if (oneOfDescriptor != null) {
+            return message.getOneofFieldDescriptor(oneOfDescriptor).getName().equals(fieldName);
+        }
+        return true;
+    }
+
     private <T> Message processFieldAfterARepeatedField(String fieldName, List<Message> repeatedMessages,
-                                                        List<T> repeatedValues, boolean lastNestedField) {
+                                                        List<T> repeatedValues, boolean lastNestedField,
+                                                        OneofDescriptor oneOfDescriptor) {
         List<Message> nextLevelRepeatedMessages = new ArrayList<>();
         Message lastProcessedMessage = null;
 
         for (Message repeatedMessage : repeatedMessages) {
             FieldDescriptor descriptor = repeatedMessage.getDescriptorForType().findFieldByName(fieldName);
+
             if (descriptor.getType().equals(FieldDescriptor.Type.MESSAGE)) {
-                // Replace upper level repeated message by next field, as we need to get all instances for this field
-                lastProcessedMessage = (Message) repeatedMessage.getField(descriptor);
-                if (lastNestedField) {
-                    repeatedValues.add(ClassUtils.cast(lastProcessedMessage));
-                } else {
-                    nextLevelRepeatedMessages.add(lastProcessedMessage);
+                // Non-Primitive type
+                if (isValidField(oneOfDescriptor, repeatedMessage, fieldName)) {
+                    lastProcessedMessage = (Message) repeatedMessage.getField(descriptor);
+
+                    if (lastNestedField) {
+                        // Case: last nested field (values to return, either primitive or non-primitives)
```

Comment:
can last field in the path be a message and not a primitive?

Reply:
yes, we can index on non-primitives or primitives